backend/backend/scopes.py

```diff
@@ -0,0 +1,93 @@
+from django.core.exceptions import ImproperlyConfigured
+from oauth2_provider.scopes import BaseScopes
+from oauth2_provider.contrib.rest_framework import TokenHasScope
+from api.models import Store
+
+
+def getPK(path):
+    elements = path.split("/")
+    for i in range(len(elements)):
+        if elements[i].upper() == "STORES":
+            if i + 1 < len(elements) and elements[i+1].isdigit():
+                return elements[i+1]
+    return ""
+
+
+class CustomScopes(BaseScopes):
+    def get_all_scopes(self):
+        scopes = {"read": "custom reading scope", "write": "custom writing scope"}
+        for store in Store.objects.all():
+            scopes["store_" + str(store.pk) + ":employee"] = "Employee scope for " + store.name
+            scopes["store_" + str(store.pk) + ":manager"] = "Manager scope for " + store.name
+            scopes["store_" + str(store.pk) + ":vendor"] = "Vendor scope for " + store.name
+        return scopes
+
+    def get_available_scopes(self, application=None, request=None, *args, **kwargs):
+        scopes = ["read", "write"]
+        for store in Store.objects.all():
+            scopes.append("store_" + str(store.pk) + ":employee")
+            scopes.append("store_" + str(store.pk) + ":manager")
+            scopes.append("store_" + str(store.pk) + ":vendor")
+        return scopes
+
+    def get_default_scopes(self, application=None, request=None, *args, **kwargs):
+        scopes = ["read", "write"]
+        for store in Store.objects.all():
+            scopes.append("store_" + str(store.pk) + ":employee")
+            scopes.append("store_" + str(store.pk) + ":manager")
+            scopes.append("store_" + str(store.pk) + ":vendor")
+        return scopes
```

Comment:
Will these methods be different in the future? @MattKersey

Reply:
Not that I can tell, this structure is just following what I found in the oauth toolkit docs, which has identical methods for the two. I can simplify and call a common method.